### ISSUE TITLE
Add participant features from precedence

### ISSUE THREAD

**Issue by MihaiSurdeanu** [clulab/reach#178]
- We only look at immediate precedence, for within sentence. Use both the deterministic and the ML model.
- These features become not_features when the preceding relation is negated!

**Comment by myedibleenso** [comment]
### A few questions...
1. How should this be stored in the Fries output?  Do we know yet?  
2. Are we only reporting features for input to an event, or both I and O?
3. Are we just storing some strings for each (ex. Phosphorylation)?
4. What about negative evidence (ex. "X is NOT phosphorylated")?  Should that be represented in the output?  Should it just be skipped?

**Comment by myedibleenso** [comment]
MITRE's specifications include these fields pertaining to features on participants:
```json
"features": [],
"not_features": []
```

### Types of participant features:
- binding features

```json
 {
   "feature_type": "binding_feature",
   "bound_to": {
       "entity_type": "protein",
       "entity_text": [
           "KRAS"
       ],
       "identifier": "UNIPROT:RASK_HUMAN"
   }
 }
```
- location features
```json
{
   "feature_type": "binding_feature",
   "bound_to": {
       "entity_type": "protein",
       "entity_text": [
           "KRAS"
       ],
       "identifier": "UNIPROT:RASK_HUMAN"
   }
}
```
- modification features

```json
{
    "feature_type": "modification_feature",
    "modification_type": "<modification type>",
    "location": "<amino acid residue number>",
    "aa_code": "<amino acid single letter code>"
}
```

Modifications features are likely to be the most common.  This is the list of valid mod features they provided:

- phosphorylation
- acetylation
- farnesylation
- glycosylation
- hydroxylation
- methylation
- ribosylation
- sumoylation
- ubiquitination

**Comment by myedibleenso** [comment]
I'm a little confused on `participant_a` vs `participant_b`.  Can someone clarify?

It seems like someone will want to know the features on "participants" going into a reaction and features on "participants" coming out of a reaction...right?

**Comment by myedibleenso** [comment]
So we need to spit out participant features, but for Fries use are we spitting them on the stuff going into a reaction as well as the stuff coming out of a reaction?

**Comment by MihaiSurdeanu** [comment]
Hi Gus,
Here's my take on these issues:
1. We only need to explicitly list these for input participants. *Not* for
outputs, which are not listed in the Fries format anyway.
2. Both participant_a and participant_b are inputs. For simple events,
participant_a is typically the controller of the regulation dominating it.
For Binding events, _a and _b are members of the complex.
3. The participant features must capture if they were generated through a
simple event (e.g., phospho) or are part of a complex.
4. We need to store more than a string for these features, but the format
should not be too complicated. We should aim to follow what MITRE has in
the spec, in terms of content. That is:
- for simple events we need to store the site as well, if available.
- for binding events, we need the other members of the complex (note that
now we may have complexes with more than 2 participants).
- for all, we need to store if they are negated or not.

Hope this helps.
Mihai

**Comment by myedibleenso** [comment]
I think I've got it now.  Thanks for clarifying!

**Comment by hickst** [comment]
Despite discussions of "input" and "output", it seems like we are still representing a "static snapshot" of an event. At some point are we going to need to represent a change of state over time?

**Comment by myedibleenso** [comment]
Not for this evaluation anyway.  But you can think of our representations of those different reactions as I-O "snapshots" in the process that together constitute your time series, @hickst.

**Comment by hickst** [comment]
MS says we should add these features to arguments of events using existing feature structures.

**Comment by myedibleenso** [comment]
I've "tested" my solution on five papers from the dry run set and they were processed without error.  I've added a method (`getInputFeaturesForParticipants`) to the `Assembler` (an assembly manager used for export).  This produces a `Seq[RoleWithFeatures]`.   

`RoleWithFeatures` is just a case class for storing an event's participant (a `Mention`), the participant's role (theme, controlled, etc.), and the input features for the participant.

```
/**
  * Storage class for an event's participant (+ the input features for the participant)
  * @param role the participant's role (theme, controlled, etc.)
  * @param mention the mention associated with the participant
  * @param features Set[[PTM]] defining the state of the participant upon event input (pulled from causal predecessors)
  */
case class RoleWithFeatures(role: String, mention: Mention, features: Set[representations.PTM])
```

The logic of finding an Event's participant features is handled by a new class, `ParticipantFeatureTracker`.  I'm not in love with the solution, but it will work for the eval...

@hickst, each member of the `Set[representations.PTMs]` accessed via `RoleWithFeatures.features` will need to be transformed into some string representation within the exporter.

This has all been pushed to the`participant-features` branch.

**Comment by hickst** [comment]
@myedibleenso Bad news: latest master merged into participant-features. Then I tried to run on DR3 files. Perhaps I made a bad call to Assembler...
```
[error] (run-main-0) scala.MatchError: org.clulab.assembly.representations.Activation@7e66ddc6 (of class org.clulab.assembly.representations.Activation)
scala.MatchError: org.clulab.assembly.representations.Activation@7e66ddc6 (of class org.clulab.assembly.representations.Activation)
	at org.clulab.assembly.ParticipantFeatureTracker.getOutputFeatures(ParticipantFeatureTracker.scala:27)
	at org.clulab.assembly.ParticipantFeatureTracker$$anonfun$getInputFeatures$1.apply(ParticipantFeatureTracker.scala:59)
	at org.clulab.assembly.ParticipantFeatureTracker$$anonfun$getInputFeatures$1.apply(ParticipantFeatureTracker.scala:59)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
	at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:241)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
	at org.clulab.assembly.ParticipantFeatureTracker.getInputFeatures(ParticipantFeatureTracker.scala:59)
	at org.clulab.assembly.Assembler$$anonfun$7$$anonfun$apply$8.apply(Assembler.scala:49)
	at org.clulab.assembly.Assembler$$anonfun$7$$anonfun$apply$8.apply(Assembler.scala:48)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:285)
	at org.clulab.assembly.Assembler$$anonfun$7.apply(Assembler.scala:48)
	at org.clulab.assembly.Assembler$$anonfun$7.apply(Assembler.scala:47)
	at scala.collection.TraversableLike$WithFilter$$anonfun$flatMap$2.apply(TraversableLike.scala:714)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
	at scala.collection.TraversableLike$WithFilter.flatMap(TraversableLike.scala:713)
	at org.clulab.assembly.Assembler.getInputFeaturesForParticipants(Assembler.scala:47)
	at org.clulab.reach.extern.export.fries.FriesOutput.org$clulab$reach$extern$export$fries$FriesOutput$$mkEventMention(FriesOutput.scala:576)
	at org.clulab.reach.extern.export.fries.FriesOutput$$anonfun$eventsToModel$1.apply(FriesOutput.scala:267)
	at org.clulab.reach.extern.export.fries.FriesOutput$$anonfun$eventsToModel$1.apply(FriesOutput.scala:264)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.clulab.reach.extern.export.fries.FriesOutput.eventsToModel(FriesOutput.scala:264)
	at org.clulab.reach.extern.export.fries.FriesOutput.writeJSON(FriesOutput.scala:171)
	at org.clulab.reach.AssemblyCLI.outputMentions(AssemblyCLI.scala:89)
	at org.clulab.reach.AssemblyCLI$$anonfun$processPapers$1.apply(AssemblyCLI.scala:65)
	at org.clulab.reach.AssemblyCLI$$anonfun$processPapers$1.apply(AssemblyCLI.scala:42)
```

**Comment by myedibleenso** [comment]
I'll work on it a bit more.  Thanks.

**Comment by hickst** [comment]
May be sick of this but it isn't done yet: output not correct and tests needed.

**Comment by hickst** [comment]
Finally closed by pull request #292
